### PR TITLE
L2 residency control helper functions

### PR DESCRIPTION
**Pull request by sleeepyjack**
This PR adds helper functions that allow for pinning a region in global memory into L2$.

Cherry-picked from #101 so it can be reviewed separately.

Also related: https://github.com/NVIDIA/cuCollections/pull/101#issuecomment-1201504489

### CODE REVIEW

File: include/cuco/detail/cache_residency_control.cuh [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ * @param[in] end End of the memory region
+ * @param[in] hit_rate Probability for a sub-segment to be mapped in L2
+ * @param[in] carve_out Fraction of total L2 space to be blocked for resident memory segments
+ *
+ * @note Only has effect on Ampere and above.
+ * @note Assumes the memory region to be contiguous.
+ */
+template <typename Iterator>
+void register_l2_persistence(
+  cudaStream_t& stream, Iterator begin, Iterator end, float hit_rate = 0.6f, float carve_out = 1.0f)
```
Comment (sleeepyjack) [inline include/cuco/detail/cache_residency_control.cuh]: For now we use an iterator pair and assume they represent a continuous memory segment. This is going to look a lot cleaner once we can use span.

File: include/cuco/detail/cache_residency_control.cuh [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ */
+void unregister_l2_persistence(cudaStream_t& stream)
+{
+  cudaStreamAttrValue stream_attribute;
+  // Setting the window size to 0 to disable it
+  stream_attribute.accessPolicyWindow.num_bytes = 0;
+  // Overwrite the access policy attribute of CUDA Stream
+  cudaStreamSetAttribute(stream, cudaStreamAttributeAccessPolicyWindow, &stream_attribute);
+  // Remove any persistent lines in L2$
+  cudaCtxResetPersistingL2Cache();
```
Comment (sleeepyjack) [inline include/cuco/detail/cache_residency_control.cuh]: This resets the cache globally. Maybe undesirable.

**Comment by sleeepyjack** [comment]
A usage example can be found here (#101): https://github.com/NVIDIA/cuCollections/blob/6b61279f8870142e81875f8c7a1fbdea04d619b4/examples/bloom_filter/l2_residency_example.cu

File: include/cuco/detail/cache_residency_control.cuh (hunk 1 above)
Comment (jrhemstad) [reply include/cuco/detail/cache_residency_control.cuh]: Can you include a description of why this is necessary vs. `annotated_ptr`?

File: include/cuco/detail/cache_residency_control.cuh [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+#include <iterator>
+
+namespace cuco {
+
+/**
+ * @brief Registers the global memory region `[begin, end)` to
+ * be permanently resident in L2 cache.
+ *
+ * @tparam Iterator Accessor of the memory region
+ * @param[in, out] stream The CUDA stream this region is accessed through
```
Comment (PointKernel) [inline include/cuco/detail/cache_residency_control.cuh]:
```suggestion
 * @tparam Iterator Accessor of the memory region
 *
 * @param[in, out] stream The CUDA stream this region is accessed through
```

File: include/cuco/detail/cache_residency_control.cuh (hunk 1 above)
Comment (PointKernel) [reply include/cuco/detail/cache_residency_control.cuh]: Could you place `stream` as the last parameter? Not necessary but try to be consistent with other cuco APIs.

**Review by PointKernel — COMMENTED** [review]
Should we move the file to `include/cuco` since it's exposed to the public?

File: include/cuco/detail/cache_residency_control.cuh (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/cache_residency_control.cuh]: The most elegant solution to this problem would be to wrap the storage of the filter as an `annotated_ptr`. However, this first requires a solution to the `annotated_ptr` + `cuda::atomic`/`cuda::atomic_ref` problem, which is not available yet.

The next best approach would be to set the access property manually using `cuda::apply_access_property`, which I implemented in the previous version of the benchmark script: https://github.com/NVIDIA/cuCollections/blob/a1ea293ff4ce4c3bbbcd1aaecb415b75dec4c5bb/benchmarks/bloom_filter/bloom_filter_bench.cu#L199-L217

This however showed little to no performance impact. I have not been able to quite get to the bottom of why this is happening. Another disadvantage of this method is that `cuda::apply_access_property` does not allow hybrid access policies, e.g. 60% `cuda::access_property_persisting` + 40% `cuda::access_property_normal`, which showed the best performance in the benchmarks.

tl;dr This is a workaround that is flexible enough for our purposes (can also be used with other cuco map types) but should be replaced once `annotated_ptr` works as expected.

Comment (sleeepyjack) [reply include/cuco/detail/cache_residency_control.cuh]: > Could you place `stream` as the last parameter? Not necessary but try to be consistent with other cuco APIs.

Yes, I can do that for the sake of consistency. I initially put it in the front since it's an in/out parameter, i.e., the stream attributes are mutated.

However, this reminds me that I find the placement of the stream parameter in our API unfortunate. I would put the stream in front of the hashers and equality operators, since you are likely to set this one explicitly more often than the others.
It also has a shorter default value compared to the others.

Comment (sleeepyjack) [reply include/cuco/detail/cache_residency_control.cuh]: Wait, I can't put the stream at the end since there are additional defaulted parameters.

**Comment by sleeepyjack** [comment]
> Should we move the file to `include/cuco` since it's exposed to the public?

I also had this in mind (that's why I added the doxygen docs). I would say yes. However, I don't know how broad the range of applications for this functionality really is.

File: include/cuco/detail/cache_residency_control.cuh (hunk 1 above)
Comment (sicario001) [reply include/cuco/detail/cache_residency_control.cuh]: > The most elegant solution to this problem would be to wrap the storage of the filter as an `annotated_ptr`. However, this first requires a solution to the `annotated_ptr` + `cuda::atomic`/`cuda::atomic_ref` problem, which is not available yet.
> 
> The next best approach would be to set the access property manually using `cuda::apply_access_property`, which I implemented in the previous version of the benchmark script:
> 
> https://github.com/NVIDIA/cuCollections/blob/a1ea293ff4ce4c3bbbcd1aaecb415b75dec4c5bb/benchmarks/bloom_filter/bloom_filter_bench.cu#L199-L217
> 
> This however showed little to no performance impact. I have not been able to quite get to the bottom of why this is happening. Another disadvantage of this method is that `cuda::apply_access_property` does not allow hybrid access policies, e.g. 60% `cuda::access_property_persisting` + 40% `cuda::access_property_normal`, which showed the best performance in the benchmarks.
> 
> tl;dr This is a workaround that is flexible enough for our purposes (can also be used with other cuco map types) but should be replaced once `annotated_ptr` works as expected.

Hey, apologies for bringing up this old PR. I am also testing the cuda::apply_access_property API (on https://leimao.github.io/blog/CUDA-L2-Persistent-Cache/) and in general the prefetch.global.L2::evict_last PTX instruction that it internally uses, but I am not able to see any effects of this. However,  when I use the accessPolicyWindow method from the CUDA API to configure L2 persistence for the same scenario, I do observe performance improvements. Did you figure out the reason or any possible resolution for this issue? Any help would be greatly appreciated.
